aws/table_aws_s3_bucket.go

```diff
@@ -318,25 +320,64 @@ func listS3Buckets(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateDa
 	return nil, nil
 }
 
-func getS3BucketEventNotificationConfigurations(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
-	// Bucket location will be nil if getBucketLocation returned an error but
-	// was ignored through ignore_error_codes config arg
-	if h.HydrateResults["getBucketLocation"] == nil {
-		return nil, nil
+func doGetBucketRegion(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData, bucket string) (string, error) {
+	// Have we already resolved and cached the bucket name?
+	c, err := getCommonColumns(ctx, d, h)
+	if err != nil {
+		plugin.Logger(ctx).Error("aws_s3_bucket.doGetBucketRegion", "get_common_columns_error", err)
+		return "", err
+	}
+	commonColumnData := c.(*awsCommonColumnData)
+	cacheKey := "getBucketRegion/" + commonColumnData.Partition + "/" + bucket
+	if cachedData, ok := d.ConnectionManager.Cache.Get(cacheKey); ok {
+		return cachedData.(string), nil
+	}
+
+	// The most reliable way to discover the region of an S3 bucket is to make an unauthenticated HTTP HEAD request.
+	// See https://github.com/aws/aws-sdk-go/issues/356#issuecomment-132707340
+	// Using the S3 path style allows to query buckets with dots in their name.
+	// Not doing so with such buckets causes `tls: failed to verify certificate: x509: certificate is valid for *.s3.amazonaws.com, s3.amazonaws.com, not www.somedomain.com.s3.amazonaws.com (SQLSTATE HV000)` errors.
+	// FIXME: do we also want to implement non S3 path style? It may help avoiding rate limiting, but given the above limitation,
+	// it may be better to define a default Steampipe limiter once actual AWS limits are discovered.
+	resp, err := http.Head(fmt.Sprintf("https://s3.amazonaws.com/%s", bucket))
+	if err != nil {
+		plugin.Logger(ctx).Error("aws_s3_bucket.doGetBucketRegion", "http_head_error", err)
+		return "", err
 	}
 
-	name := h.Item.(types.Bucket).Name
-	location := h.HydrateResults["getBucketLocation"].(*s3.GetBucketLocationOutput)
+	// We only care about the 400 and 404 HTTP status codes which respectively mean the request is invalid or the bucket does not exist at all.
+	// FIXME: do 429 responses happen and also include the header?
+	if resp.StatusCode == 400 || resp.StatusCode == 404 {
```

Comment:
We may want to also check if the status code is >= 500, or revert the condition altogether and only accept 200, 301 and 403 as valid.

Reply:
Are 200, 301, and 403 the only status codes where we get the correct region information? Do you have example responses for those error codes and the other error codes we'd want to ignore?

Reply:
> Are 200, 301, and 403 the only status codes where we get the correct region information?

Maybe not, that part is not much documented. Now that I think about, there's probably 307 too which is frequently used by AWS APIs too.

> Do you have example responses for those error codes and the other error codes we'd want to ignore?

Sadly not.

Reply:
To avoid noise while making errors available, is it helpful if we switched the level of the log message below to `Debug` and always logged it? Or is it common to get errors back from the HTTP request, even in cases where we're getting the right info back?